Add support for ::class to constant()

dunglas opened the pull request:
This change allows retrieving the value of the `::class` special constant using the `constant()` function:

```php
var_dump(
  constant('\DateTime::class')
);
```

For instance, Twig's `constant()` helper internally uses this function. This patch allows writing code like that: 
```twig
`myObject` contains a random object, retrieve its class:
{{ constant('class', myObject) }}
```

TODO: write an RFC?

francislavoie commented:
Why not just use https://www.php.net/manual/en/function.get-class.php ?

drealecs commented on ext/standard/tests/general_functions/constant_001.phpt:
```diff
@@ -0,0 +1,12 @@
+
+class Bar {}
+
+var_dump(constant('Foo\Bar::class'));
```
Because `constant('Foo\Bar::class')` is just the same as `'Foo\Bar'` that you already have to have as string, I think this need is not really real.

The current twig implementation is:
```php
function twig_constant($constant, $object = null)
{
    if (null !== $object) {
        $constant = \get_class($object).'::'.$constant;
    }

    return \constant($constant);
}
```
And by following the logic described in the first phrase, you can just make a PR to twig and have it changed to:
```php
function twig_constant($constant, $object = null)
{
    if (null !== $object) {
        if ('class' === $constant) {
            return \get_class($object);
        }
        $constant = \get_class($object).'::'.$constant;
    }

    return \constant($constant);
}
```

dunglas replied:
Replied in https://github.com/php/php-src/pull/6763#issuecomment-794995221

dunglas commented:
@drealecs @francislavoie Sure I could patch Twig (and I will), but my goal here is to make the languages consistent. `constant()` allows to access to all class constants except this one, that's weird.

dunglas commented:
Thanks for the feedback @claudepache!
I added tests covering the edge cases you mentioned (and for `defined()`, which is also supported).

I also updated the implementation to not trigger autoloading and not complain about missing classes, but my patch introduces a memory leak (https://github.com/php/php-src/pull/6763/files#diff-6f9b1b4ec76510c10bf1c606e8629eb6b4d5f9df8f21258fd73a3e76bcab4796R384) and I don't know how to fix it. The problem is that for non-existing classes we must create a new `zval` containing the class name, but the destructor of the  returned`zval` is never called by parent functions because they expect that we return an existing `zval`, not a new one. Regardless of if we merge this patch or not, I would appreciate some guidance on how to fix this, for my curiosity and to learn more about PHP internals :)

> You should consider (and justify) whether or not you add support in ReflectionClassConstant.

This is a good one! On the one hand, adding support in `ReflectionClassConstant` would also make `::class` less special and would increase the consistency of the language, but on the other hand, it can be annoying if `RefelectionClass::getConstants()` starts returning `class` in the list: this would be a BC break (and this wouldn't be very practical to have to always filter this constant). Maybe could we "hide" `class` behind a new filter, or support it when calling `ReflectionClassConstan()` or `RefelectionClass::getConstant()` directly but not include `class` in the return of `getConstants()`.

As a side effect, adding `class` to `ce->constants_table` could fix my memory leak issue, but would consume more memory for an edge case.

beberlei commented:
The larger question would be what to do abut magic constants, for example `constant("__DIR__")` et all also not work.

nikic commented:
> Thanks for the feedback @claudepache!
> I added tests covering the edge cases you mentioned (and for `defined()`, which is also supported).
> 
> I also updated the implementation to not trigger autoloading and not complain about missing classes, but my patch introduces a memory leak (https://github.com/php/php-src/pull/6763/files#diff-6f9b1b4ec76510c10bf1c606e8629eb6b4d5f9df8f21258fd73a3e76bcab4796R384) and I don't know how to fix it. The problem is that for non-existing classes we must create a new `zval` containing the class name, but the destructor of the returned`zval` is never called by parent functions because they expect that we return an existing `zval`, not a new one. Regardless of if we merge this patch or not, I would appreciate some guidance on how to fix this, for my curiosity and to learn more about PHP internals :)

It's not possible to fix this without changing the API. You'd convert `zval *zend_get_constant_ex(zend_string *cname, ...)` into `bool zend_get_constant_ex(zval *result, zend_string *cname, ...)` with the caller responsible for releasing the result.

dunglas commented:
Thanks, @nikic! That's what I thought. An API change like this is probably too big to support such an edge case.

nikic commented:
> Thanks, @nikic! That's what I thought. An API change like this is probably too big to support such an edge case.

Changing the API isn't really a big deal. It's not used much and largely for internal use.

For reference, the internals discussion for this change: https://externals.io/message/113426 As the reception wasn't super positive, do you plan to pursue an RFC on the topic?

dunglas commented:
Thanks for the explanation regarding how to change the API!

> do you plan to pursue an RFC on the topic?

As there is no consensus on this and it provides only small benefits, no I don't think so. Let's close!
